src/transformers/modeling_longformer.py

```diff
@@ -0,0 +1,667 @@
+        x_padded = F.pad(x, direction, value=padding_value)
+        x_padded = x_padded.view(*x_padded.size()[:-2], x_padded.size(-1), x_padded.size(-2))
+        return x_padded
+
+    def _skew2(self, x, padding_value):
+        '''shift every row 1 step to right converting columns into diagonals'''
+        # X = B x C x M x L
+        B, C, M, L = x.size()
+        x = F.pad(x, (0, M + 1), value=padding_value)  # B x C x M x (L+M+1)
+        x = x.view(B, C, -1)  # B x C x ML+MM+M
+        x = x[:, :, :-M]  # B x C x ML+MM
+        x = x.view(B, C, M, M + L)  # B x C, M x L+M
+        x = x[:, :, :, :-1]
+        return x
+
+    def _chunk(self, x, w):
+        '''convert into overlapping chunkings. Chunk size = 2w, overlap size = w'''
+
+        # non-overlapping chunks of size = 2w
+        x = x.view(x.size(0), x.size(1) // (w * 2), w * 2, x.size(2))
+
+        # use `as_strided` to make the chunks overlap with an overlap size = w
+        chunk_size = list(x.size())
+        chunk_size[1] = chunk_size[1] * 2 - 1
+
+        chunk_stride = list(x.stride())
+        chunk_stride[1] = chunk_stride[1] // 2
+        return x.as_strided(size=chunk_size, stride=chunk_stride)
+
+    @lru_cache()
+    def _get_invalid_locations_mask(self, w, device):
+        affected_seq_len = w
+        diagonals_list = []
+        for j in range(-w, 1, 1):
+            diagonal_mask = torch.zeros(affected_seq_len, device='cpu', dtype=torch.uint8)
+            diagonal_mask[:-j] = 1
+            diagonals_list.append(diagonal_mask)
+        mask = torch.stack(diagonals_list, dim=-1)
+        mask = mask[None, :, None, :]
+        ending_mask = mask.flip(dims=(1, 3)).bool().to(device)
+        return affected_seq_len, mask.bool().to(device), ending_mask
+
+    def _mask_invalid_locations(self, input_tensor, w) -> torch.Tensor:
+        # TODO: replace this function and `_get_invalid_locations_mask` with a simpler one that doesn't require lru_cache
+        affected_seq_len, beginning_mask, ending_mask = self._get_invalid_locations_mask(w, input_tensor.device)
+        seq_len = input_tensor.size(1)
+        beginning_input = input_tensor[:, :affected_seq_len, :, :w + 1]
+        beginning_mask = beginning_mask[:, :seq_len].expand(beginning_input.size())
+        beginning_input.masked_fill_(beginning_mask, -float('inf'))
+        ending_input = input_tensor[:, -affected_seq_len:, :, -(w + 1):]
+        ending_mask = ending_mask[:, -seq_len:].expand(ending_input.size())
+        ending_input.masked_fill_(ending_mask, -float('inf'))
+
+    def _sliding_chunks_matmul_qk(self, q: torch.Tensor, k: torch.Tensor, w: int, padding_value: float):
+        '''Matrix multiplicatio of query x key tensors using with a sliding window attention pattern.
+        This implementation splits the input into overlapping chunks of size 2w (e.g. 512 for pretrained Longformer)
+        with an overlap of size w'''
+        bsz, seqlen, num_heads, head_dim = q.size()
```

Comment:
(nit) I like it better to write out `batch_size` here since this is the usual way we write it in the library.